### PR TITLE
Update go mod to use go1.21 and update compiler switches/ flags for static linking

### PR DESCRIPTION
**Pull request by vedantjoshi84**

### CODE REVIEW

**Review by cruizen — CHANGES_REQUESTED** [review]
Hi @vedantjoshi84 , I was talking to @mithilarun  that the `make build` target looks right with the additional ldflags but we somehow still don't have a statically linked executable. He brought to my notice that the Teamcity build calls the script `build.bash` that calls the target `make container-build`. This target uses `golang:1.18` and more importantly doesn't use the same flags (`CGO_ENABLED=0` and the ldflags added recently).

File: Makefile [hunk 1]
```diff
@@ -29,7 +29,7 @@ GOFLAGS ?= ""
 default: $(BIN)
 
 container-build:
-	docker run --rm --env XDG_CACHE_HOME=$(XDG_CACHE_HOME) --env SEGMENT_KEY_PRD_PMKFT=$(SEGMENT_KEY_PRD_PMKFT) --env VERSION_OVERRIDE=${VERSION_OVERRIDE} --env GOPATH=/tmp --env GOFLAGS=$(GOFLAGS) --user $(CONT_USER):$(CONT_GRP) --volume $(PWD):$(PACKAGE_GOPATH) $(GIT_STORAGE_MOUNT) --workdir $(PACKAGE_GOPATH) golang:1.18 make
+	docker run --rm --env XDG_CACHE_HOME=$(XDG_CACHE_HOME) --env SEGMENT_KEY_PRD_PMKFT=$(SEGMENT_KEY_PRD_PMKFT) --env VERSION_OVERRIDE=${VERSION_OVERRIDE} --env GOPATH=/tmp --env GOFLAGS=$(GOFLAGS) --user $(CONT_USER):$(CONT_GRP) --volume $(PWD):$(PACKAGE_GOPATH) $(GIT_STORAGE_MOUNT) --workdir $(PACKAGE_GOPATH) golang:1.18 make LDFLAGS="-s -w -extldflags '-static'"
```
Comment (cruizen) [inline Makefile:32]:
```suggestion
	docker run --rm --env XDG_CACHE_HOME=$(XDG_CACHE_HOME) --env SEGMENT_KEY_PRD_PMKFT=$(SEGMENT_KEY_PRD_PMKFT) --env VERSION_OVERRIDE=${VERSION_OVERRIDE} --env GOPATH=/tmp --env GOFLAGS=$(GOFLAGS) --user $(CONT_USER):$(CONT_GRP) --volume $(PWD):$(PACKAGE_GOPATH) $(GIT_STORAGE_MOUNT) --workdir $(PACKAGE_GOPATH) golang:1.18 make ldflags="-s -w -extldflags '-static'"
```

Comment (cruizen) [reply Makefile:32]: `LDFLAGS` is lowercase in the build target. Uppercase here.

**Review by cruizen — CHANGES_REQUESTED** [review]

**Review by mithilarun — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (cruizen) [reply Makefile]: Can you pass ldflags

**Review by cruizen — COMMENTED** [review]
Suggested a change in `container-build`

**Pull request merged**